### PR TITLE
HPCC-14205 Add paging functions to ws_access

### PR DESCRIPTION
**Pull request by wangkx**
Four new APIs are added: UserQuery, GroupQuery, and
ResourceQuery, and GroupMemberQuery.

Signed-off-by: wangkx kevin.wang@lexisnexis.com

### CODE REVIEW

**Comment by wangkx** [comment]
@RussWhitehead please review.

**Comment by wangkx** [comment]
I made the changes based on your comments. Please review again.

**Comment by RussWhitehead** [comment]
Looks good

**Comment by wangkx** [comment]
@richardkchapman this PR is ready to merge.

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 1]
```diff
@@ -3168,6 +3325,95 @@ class CLdapClient : public CInterface, implements ILdapClient
...
+            IMPLEMENT_IINTERFACE_USING(CSimpleInterface);
+
+            CElementsPager(ILdapClient* _ldapClient, SecResourceType _rtype, const char * _basedn, const char* _resourceName,
+                unsigned _extraNameFilter, const char*_sortOrder) : ldapClient(_ldapClient), rtype(_rtype), basedn(_basedn),
+                resourceName(_resourceName), extraNameFilter(_extraNameFilter), sortOrder(_sortOrder) { };
+            virtual IRemoteConnection* getElements(IArrayOf<IPropertyTree>& elements)
+            {
+                StringArray unknownAttributes;
+                Owned<IPropertyTreeIterator> iter = ldapClient->getResourceIterator(rtype, basedn.get(), "", resourceName.get(), extraNameFilter);
+                sortElements(iter, sortOrder.get(), NULL, NULL, unknownAttributes, elements);
```
Comment (richardkchapman) [inline system/security/LdapSecurity/ldapconnection.cpp]: I don't really understand this code but is this always sorting locally having paged through all the values retrieved form ldap? If so doesn't it defeat the purpose of paging the info from ldap?

Comment (RussWhitehead) [inline system/security/LdapSecurity/ldapconnection.cpp]: I will let Kevin elaborate but all the results that come back from LDAP are unsorted, so you really need to retrieve all the pages before a sort makes sense.

Comment (wangkx) [inline system/security/LdapSecurity/ldapconnection.cpp]: Since LDAP cannot return sorted results, ESP has to sort locally.

**Pull request merged**